ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -0,0 +1,137 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota.capacityunit;
+
+import com.github.ambry.config.QuotaConfig;
+import com.github.ambry.messageformat.BlobInfo;
+import com.github.ambry.quota.QuotaEnforcer;
+import com.github.ambry.quota.QuotaException;
+import com.github.ambry.quota.QuotaName;
+import com.github.ambry.quota.QuotaRecommendation;
+import com.github.ambry.quota.QuotaResource;
+import com.github.ambry.quota.QuotaSource;
+import com.github.ambry.quota.QuotaUtils;
+import com.github.ambry.rest.RestRequest;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.quota.QuotaUtils.*;
+
+
+/**
+ * Implementation of {@link QuotaEnforcer} for Capacity Units of Ambry resource.
+ */
+public class AmbryCUQuotaEnforcer implements QuotaEnforcer {
+  private static final Logger logger = LoggerFactory.getLogger(AmbryCUQuotaEnforcer.class);
+  private static final long THROTTLE_RETRY_AFTER_MS = 1;
+  private static final long NO_THROTTLE_RETRY_AFTER_MS = -1;
+  private static final int NO_THROTTLE_HTTP_STATUS = 200;
+  private static final int THROTTLE_HTTP_STATUS = 429;
+
+  private final QuotaSource quotaSource;
+  private final QuotaConfig quotaConfig;
+
+  /**
+   * Constructor for {@link AmbryCUQuotaEnforcer}.
+   * @param quotaSource {@link QuotaSource} where the quota limit and usage will be saved and retrieved from.
+   * @param quotaConfig {@link QuotaConfig} object.
+   */
+  public AmbryCUQuotaEnforcer(QuotaSource quotaSource, QuotaConfig quotaConfig) {
+    this.quotaSource = quotaSource;
+    this.quotaConfig = quotaConfig;
+  }
+
+  @Override
+  public void init() throws QuotaException {
+    quotaSource.init();
+  }
+
+  @Override
+  public QuotaRecommendation chargeAndRecommend(RestRequest restRequest, BlobInfo blobInfo,
+      Map<QuotaName, Double> requestCostMap) throws QuotaException {
+    try {
+      if (requestCostMap.size() == 0) {
+        String errorMessage = String.format("Empty cost map provided for %s request for blob %s. Nothing to charge",
+            (restRequest == null ? "null" : restRequest.getRestMethod().name()),
+            (blobInfo == null ? "null" : blobInfo.getBlobProperties()));
+        logger.warn(errorMessage);
+        throw new QuotaException(errorMessage, true);
+      }
+      QuotaName quotaName = getCUQuotaName(restRequest);
+      if(!quotaSource.isReady()) {
+        logger.warn("Allowing request to go through as CU QuotaSource is not ready.");
+        return new QuotaRecommendation(false, 0, quotaName, 200, -1);
+      }
+      quotaSource.chargeUsage(getQuotaResource(restRequest), quotaName, requestCostMap.get(quotaName));
```

Comment:
Note that QuotaEnforcer is not designed to be thread safety. Thread safety will come with the new QuotaManager implementation. The scope of this PR is to introduce the new QuotaEnforcer and QuotaSource for CapacityUnit quota, and make changes in other code so that the overall behavior remains unchanged.